### PR TITLE
[Documentation] Robust state tomography tutorial

### LINKED ISSUE #51: [Documentation] Add example of robust state tomography
Add a page in the doc showing an example of state tomography comparing robust shadows and vanilla shadows.

### PR DESCRIPTION
**Pull request by chMoussa**
Closes #51. Also using a different zero state calibration procedure.

### CODE REVIEW

File: docs/measurements/index.md [hunk 1]
```diff
@@ -19,6 +21,7 @@ $$
 \langle \hat{\mathcal{P}} \rangle=\sum_{z,z'}\langle z |\bar{a}_z\hat{\mathcal{D}}a_{z'}|z'\rangle = \sum_{z}|a_z|^2(-1)^{\phi_z(\hat{\mathcal{P}})}
 $$
 
+### Using tomography
```
Comment (RolandMacDoland) [inline docs/measurements/index.md:24]: Is it necessary to have one more hierarchical level for such a small paragraph ?

File: docs/measurements/index.md [hunk 2]
```diff
@@ -59,7 +62,7 @@ print(f"Exact expectation value = {exact_values}") # markdown-exec: hide
 print(f"Estimated expectation value tomo = {estimated_values_tomo}") # markdown-exec: hide
 ```
 
-## Getting measurements
+### Getting measurements
```
Comment (RolandMacDoland) [inline docs/measurements/index.md:65]: Same.

File: docs/measurements/index.md (hunk 1 above)
Comment (RolandMacDoland) [reply docs/measurements/index.md:24]: Ah sorry I had to expand to see the whole text.

File: docs/measurements/index.md [hunk 3]
```diff
@@ -141,27 +146,12 @@ print(exp_reconstructed_state) # markdown-exec: hide
...
 
-Robust shadows [^4] were built upon the classical shadow scheme but have the particularity to be noise-resilient. Using an experimentally friendly calibration procedure, one can eﬃciently characterize and mitigate noises in the shadow estimation scheme, given only minimal assumptions on the experimental conditions. Such a procedure has been used in [^5] to estimate the Quantum Fisher information out of a quantum system. Note that robust shadows are equivalent to classical shadows in non-noisy settings by setting the `calibration` coefficients to $\frac{1}{3}$ for each qubit.
+Robust shadows [^4] were built upon the classical shadow scheme but have the particularity to be noise-resilient. Using an experimentally friendly calibration procedure, one can eﬃciently characterize and mitigate noises in the shadow estimation scheme, given only minimal assumptions on the experimental conditions. Such a procedure has been used in [^5] to estimate the Quantum Fisher information out of a quantum system. Note that robust shadows are equivalent to classical shadows in non-noisy settings by setting the `calibration` coefficients to $\frac{1}{3}$ for each qubit, as shown below:
```
Comment (RolandMacDoland) [inline docs/measurements/index.md]: What is "experimentally friendly" ?

File: docs/measurements/shadow_tomography.md [hunk 4]
```diff
@@ -0,0 +1,138 @@
+# Robust shadow tomography
+
+In this tutorial, we will estimate a physical property out of a quantum system, namely the purity of the partial traces, in the presence of measurement noise. To do so, we will use the formalism of classical shadows[^1], and especially their robust version[^2]. This tutorial is inspired from a notebook example of robust shadow tomography from the randomized measurements toolbox in julia[^3].
```
Comment (RolandMacDoland) [inline docs/measurements/shadow_tomography.md]:
```suggestion
In this tutorial, we will estimate a physical property out of a quantum system, namely the purity of the partial traces, in the presence of measurement noise. To do so, we will use the formalism of classical shadows[^1], and especially their robust version[^2]. This tutorial is inspired from a notebook example of robust shadow tomography from the randomized measurements toolbox in Julia[^3].
```

Maybe cite the authors ?

File: docs/measurements/shadow_tomography.md [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+
+
+## Setting the model
+
+First, we will set the noise model and a circuit from which we will estimate the purity.
+
+### Noise model
+We will use a depolarizing noise model with a different error probability per qubit.
```
Comment (RolandMacDoland) [inline docs/measurements/shadow_tomography.md]:
```suggestion
### Noise model

We will use a depolarizing noise model with a different error probability per qubit.
```

File: docs/measurements/shadow_tomography.md [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+
+```python exec="on" source="material-block" session="shadow_tomo" result="json"
+import torch
+from qadence import NoiseHandler, NoiseProtocol
+
+torch.manual_seed(0)
+nqbits = 2
```
Comment (RolandMacDoland) [inline docs/measurements/shadow_tomography.md]:
```suggestion
n_qubits = 2
```

File: docs/measurements/shadow_tomography.md [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+values = {
+    "theta1": torch.tensor([0.5]),
+    "theta2": torch.tensor([1.5]),
+    "theta3": torch.tensor([2.0]),
+    "theta4": torch.tensor([2.5]),
+}
+
+model = QuantumModel(
+    circuit=circuit,
+    observable=[], # no observable needed here
```
Comment (RolandMacDoland) [inline docs/measurements/shadow_tomography.md]: Is it necessary ? Can we simply delete that line and add the comment at the top ?

File: docs/measurements/shadow_tomography.md [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+
+For calculating purities, we can use the utility functions `apply_partial_trace` and `compute_purity`:
+
+```python exec="on" source="material-block" session="shadow_tomo" result="json"
+from qadence_protocols.utils_trace import apply_partial_trace, compute_purity
+
+def partial_purities(density_mat):
+    purities = []
+    for i in range(nqbits):
+        partial_trace_i = apply_partial_trace(density_mat, [i]).squeeze()
```
Comment (RolandMacDoland) [inline docs/measurements/shadow_tomography.md]:
```suggestion
        partial_trace_i = partial_trace(density_mat, [i]).squeeze()
```

File: docs/measurements/shadow_tomography.md [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+        purities.append(compute_purity(partial_trace_i))
```
Comment (RolandMacDoland) [inline docs/measurements/shadow_tomography.md]:
```suggestion
        purities.append(purity(partial_trace_i))
```

File: docs/measurements/shadow_tomography.md [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+
+    return torch.tensor(purities)
+
+expected_purities = partial_purities(model.run(values))
+print(f"Expected purities = {expected_purities}") # markdown-exec: hide
+```
+
+### Add noise to circuit
+The circuit is defined as follows where we set the previous noise model in the last operations as measurement noise.
```
Comment (RolandMacDoland) [inline docs/measurements/shadow_tomography.md]:
```suggestion
### Add noise to circuit

The circuit is defined as follows where we set the previous noise model in the last operations as measurement noise.
```

File: docs/measurements/shadow_tomography.md [hunk 11]
```diff
@@ -0,0 +1,138 @@
...
+noisy_model = QuantumModel(
+    circuit=noisy_circuit,
+    observable=[], # no observable needed here
+)
+```
+
+## Shadow estimations
+
+### Vanilla classical shadows
+We will first run vanilla shadows to reconstruct the density matrix representation of the circuit, from which we can estimate the purities.
```
Comment (RolandMacDoland) [inline docs/measurements/shadow_tomography.md]:
```suggestion
### Vanilla classical shadows

We will first run vanilla shadows to reconstruct the density matrix representation of the circuit, from which we can estimate the purities.
```

File: docs/measurements/shadow_tomography.md [hunk 12]
```diff
@@ -0,0 +1,138 @@
...
+As we can see below, the estimated purities diverge from the expected ones due to the presence of noise. Next, we will use robust shadows to mitigate the noise effect.
```
Comment (RolandMacDoland) [inline docs/measurements/shadow_tomography.md]: I would put the second sentence after the code block as a transition to the next section.

File: qadence_protocols/measurements/calibration.py [hunk 13]
```diff
@@ -1,19 +1,79 @@
...
 from qadence.types import Endianness
+from torch import Tensor
+
+from qadence_protocols.measurements.utils_shadow.data_acquisition import extract_operators
+from qadence_protocols.utils_trace import apply_partial_trace
+
 
-pauli_gates = [X, Y, Z]
-pauli_tensors = [XMAT[0], YMAT[0], ZMAT[0]]
+def _get_noiseless_probas(
```
Comment (RolandMacDoland) [inline qadence_protocols/measurements/calibration.py]:
```suggestion
def _noiseless_probabilites(
```

File: qadence_protocols/measurements/calibration.py [hunk 14]
```diff
@@ -1,19 +1,79 @@
...
+
+    Args:
+        n_qubits (int): Number of qubits.
+        samples (Counter): Samples obtained from a circuit.
+        endianness (Endianness, optional): Endianness of operations. Defaults to Endianness.BIG.
+
+    Returns:
+        Tensor: _description_
+    """
+    freqs = torch.zeros((n_qubits, 2))
```
Comment (RolandMacDoland) [inline qadence_protocols/measurements/calibration.py]:
```suggestion
    frequencies = torch.zeros((n_qubits, 2))
```

File: qadence_protocols/measurements/calibration.py [hunk 15]
```diff
@@ -1,19 +1,79 @@
...
+            endianness=endianness,
+        )
+        for i in range(n_qubits):
+            noiseless_probas[r][i] = torch.diagonal(
+                apply_partial_trace(wave_fct, [i]), dim1=1, dim2=2
+            ).real.squeeze()
+    return noiseless_probas
+
+
+def _singlebit_freq_from_samples(
```
Comment (RolandMacDoland) [inline qadence_protocols/measurements/calibration.py]:
```suggestion
def _samples_frequencies(
```

File: qadence_protocols/measurements/calibration.py [hunk 16]
```diff
@@ -46,45 +106,54 @@ def zero_state_calibration(
...
+            for proto, options in zip(digital_part.protocol, digital_part.options):
+                target = options.get("target", None)
+                if target is not None:
+                    noisy_identities.append(I(target=target, noise=NoiseHandler(proto, options)))
+                else:
+                    for target in range(n_qubits):
+                        noisy_identities.append(
+                            I(target=target, noise=NoiseHandler(proto, options))
+                        )
+            noisy_zero_circ = QuantumCircuit(n_qubits, *noisy_identities)
```
Comment (RolandMacDoland) [inline qadence_protocols/measurements/calibration.py:127]: Couldn't all this be hidden in the `NoiseHandler` ? It feels a bit cumbersome here.

File: qadence_protocols/measurements/shadow.py [hunk 17]
```diff
@@ -69,15 +69,20 @@ def validate_options(self, options: dict) -> dict:
                 " 'shadow_size' of type 'int' or 'accuracy' of type 'float'."
             )
         confidence = options.get("confidence", None)
-        if confidence is None:
-            raise KeyError("Shadow protocol requires an option 'confidence' of type 'float'.")
+        shadow_medians = options.get("shadow_medians", None)
+        if confidence is None and shadow_medians is None:
+            raise KeyError(
+                "Shadow protocol requires either an option"
```
Comment (RolandMacDoland) [inline qadence_protocols/measurements/shadow.py]:
```suggestion
                "Shadow protocol requires either an option "
```

File: qadence_protocols/measurements/utils_shadow/data_acquisition.py [hunk 18]
```diff
@@ -189,26 +189,34 @@ def shadow_samples(
     unitary_ids = np.random.randint(0, 3, size=(shadow_size, circuit.n_qubits))
     shadow: list = list()
     all_rotations = extract_operators(unitary_ids, circuit.n_qubits)
-    all_rotations = [
-        QuantumCircuit(circuit.n_qubits, rots) if rots else QuantumCircuit(circuit.n_qubits)
-        for rots in all_rotations
-    ]
+
+    initial_state = state
+    if isinstance(backend, PyQBackend) or backend.backend.name == BackendName.PYQTORCH:
```
Comment (RolandMacDoland) [inline qadence_protocols/measurements/utils_shadow/data_acquisition.py:194]: Is it necessary to do both these tests ?

File: qadence_protocols/utils_trace.py [hunk 19]
```diff
@@ -99,3 +100,35 @@ def expectation_trace(state: Tensor, observables: list[AbstractBlock]) -> Tensor
...
+def compute_purity(rho: Tensor) -> Tensor:
+    """Compute the purity of a density matrix.
+
+    Args:
+        rho (Tensor): Density matrix.
+
+    Returns:
+        Tensor: Tr[rho ** 2]
+    """
+    return torch.trace(rho**2).real
```
Comment (RolandMacDoland) [inline qadence_protocols/utils_trace.py:134]:
```suggestion
    return torch.trace(rho ** 2).real
```

**Review by RolandMacDoland — COMMENTED** [review]
Thanks @chMoussa. As usual, renaming comments. Happy to approve soon after.

File: docs/measurements/index.md (hunk 3 above)
Comment (chMoussa) [reply docs/measurements/index.md]: It is just about measuring the 0 states, and computing individual coefficients for each qubit from the measurements.

File: docs/measurements/index.md (hunk 2 above)
Comment (chMoussa) [reply docs/measurements/index.md:65]: I like clear separations, even for small texts.

File: qadence_protocols/measurements/utils_shadow/data_acquisition.py (hunk 18 above)
Comment (chMoussa) [reply qadence_protocols/measurements/utils_shadow/data_acquisition.py:194]: It helps me reduce computation time by half, especially important for CI purposes

File: qadence_protocols/measurements/calibration.py (hunk 16 above)
Comment (chMoussa) [reply qadence_protocols/measurements/calibration.py:127]: I had to because it did not work with me. I think we will need to modify qadence for this.

File: docs/measurements/shadow_tomography.md (hunk 7 above)
Comment (chMoussa) [reply docs/measurements/shadow_tomography.md]: Yes with a slight modification in the source code to avoid calling obs.abstract when the model's observable is None

File: docs/measurements/index.md (hunk 2 above)
Comment (RolandMacDoland) [reply docs/measurements/index.md:65]: Yeah, I tend not to add to many layers in the docs but hey, that's a stylistic opinion.

File: docs/measurements/index.md (hunk 3 above)
Comment (RolandMacDoland) [reply docs/measurements/index.md]: I guess it makes more sense to add these details. "Experimentally friendly" is a bit meaningless.

File: qadence_protocols/measurements/calibration.py (hunk 16 above)
Comment (RolandMacDoland) [reply qadence_protocols/measurements/calibration.py:127]: What do you have in mind in Qadence ?

File: qadence_protocols/measurements/utils_shadow/data_acquisition.py (hunk 18 above)
Comment (RolandMacDoland) [reply qadence_protocols/measurements/utils_shadow/data_acquisition.py:194]: It's a bit weird, I don't see PyQBackend instances not being named `BackendName.PYQTORCH` ?

**Review by RolandMacDoland — APPROVED** [review]
Thanks @chMoussa. Maybe you can be a bit more explicit on some points above and raise relevant issues if needed. Otherwise LGTM.

File: qadence_protocols/measurements/utils_shadow/data_acquisition.py (hunk 18 above)
Comment (chMoussa) [reply qadence_protocols/measurements/utils_shadow/data_acquisition.py:194]: I think more like the second part tackles differentiablebackends

File: qadence_protocols/measurements/calibration.py (hunk 16 above)
Comment (chMoussa) [reply qadence_protocols/measurements/calibration.py:127]: Checking at least when applying `NoiseHandler` where options are defined with `target`, that only operators acting on `target` are affected. But that's to be well tested. Also that empty circuits can be handled.
